# 🤖 Agentic AI Capstone Project
## Course Assistant — Agentic AI Hands-On Course 2026

**Name:** Prawesh yadav   
**Roll Number:** 23053896
**Batch/Program:** Agentic AI Course 2026  
**Instructor:** Dr. Kanthi Kiran Sirra  

---

## Problem Statement
**Domain:** Agentic AI Course Content  
**User:** B.Tech students enrolled in the Agentic AI Hands-On Course  
**Problem:** Students need concept help at odd hours. They struggle to find answers about LangGraph, ChromaDB, RAG, MemorySaver, RAGAS metrics, Streamlit deployment, and other course topics without waiting for the instructor.  
**Success:** An assistant that answers course-related questions faithfully from the course knowledge base, remembers the conversation, uses tools for date/math/web queries, and never hallucates — achieving a RAGAS faithfulness score above 0.7.  
**Tool:** DateTime (current date), Calculator (math), Web Search (live info)


In [1]:
# Run this cell first to install all required packages
import subprocess, sys
packages = [
    'langgraph', 'langchain-groq', 'langchain-core',
    'sentence-transformers', 'chromadb',
    'streamlit', 'ragas', 'datasets'
]
for pkg in packages:
    subprocess.check_call([sys.executable, '-m', 'pip', 'install', '-q', pkg])
print('✅ All packages installed!')

✅ All packages installed!


## Set Your Groq API Key

In [2]:
import os
from dotenv import load_dotenv

load_dotenv(dotenv_path="project/.env")

print("groq api key load successfully:")

groq api key load successfully:


## Part 1 — Knowledge Base Setup

In [3]:
from knowledge_base import DOCUMENTS
print(f'Total documents: {len(DOCUMENTS)}')
for doc in DOCUMENTS:
    print(f"  {doc['id']}: {doc['topic']}")

Total documents: 13
  doc_001: What is Agentic AI
  doc_002: LangGraph and StateGraph
  doc_003: ChromaDB and RAG
  doc_004: Memory and MemorySaver
  doc_005: Router Node and Routing Logic
  doc_006: Answer Node and System Prompt Design
  doc_007: Eval Node and Self-Reflection
  doc_008: Tool Use — DateTime, Calculator, Web Search
  doc_009: RAGAS Evaluation Metrics
  doc_010: Streamlit Deployment
  doc_011: Red-Teaming and Adversarial Testing
  doc_012: Course Structure — 13 Days Overview
  doc_013: What is LangChain


In [4]:
import sys
!{sys.executable} -m pip install tavily-python


[notice] A new release of pip is available: 25.0.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [5]:
import os
from tavily import TavilyClient

client = TavilyClient(api_key=os.getenv("TAVILY_API_KEY"))
print("Tavily client initialized successfully!")

Tavily client initialized successfully!


## Tavily api key

In [6]:
import os
from dotenv import load_dotenv
from tavily import TavilyClient

load_dotenv()

client = TavilyClient(api_key=os.getenv("TAVILY_API_KEY"))

print(client.search("who is elon musk"))

{'query': 'who is elon musk', 'response_time': 0.77, 'follow_up_questions': None, 'answer': None, 'images': [], 'results': [{'url': 'https://www.edge.org/memberbio/elon_musk', 'title': 'Elon Musk - Edge.org', 'content': 'https://www.edge.org/memberbio/elon\\_musk. # Elon Musk. Engineer & Entrepreneur; CEO & CTO, SpaceX; CEO, Head of Product Design, Tesla, Inc. ELON MUSK is an\xa0entrepreneur and\xa0business magnate who builds and operates companies to solve environmental, social, and economic challenges. In 2002, Musk founded SpaceX, an\xa0aerospace\xa0manufacturer and space transport services company, of which he is CEO and CTO. Musk’s achievements in business and technology have been widely recognized. Based on his work at SpaceX, Tesla Motors, and SolarCity, *Research and Development Magazine* named Musk its innovator of the year, and *Inc.* named him its entrepreneur of the year. Musk is a trustee of the X Prize Foundation, which promotes renewable energy technologies, and sits on 

## groq api key

In [7]:
import os
from dotenv import load_dotenv

load_dotenv()

if os.getenv("GROQ_API_KEY"):
    print("GROQ API key is set successfully!")
    

GROQ API key is set successfully!


In [8]:
import importlib
import agent

importlib.reload(agent)

from agent import build_knowledge_base

embedder, collection = build_knowledge_base()

print("✅ Knowledge base loaded successfully!")

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.


✅ Knowledge base loaded successfully!


In [9]:
import importlib
import agent

importlib.reload(agent)

from agent import build_knowledge_base, test_retrieval

embedder, collection = build_knowledge_base()
test_retrieval(embedder, collection)

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.



🔍 Retrieval Test:

Result 1:

LangChain is a powerful framework used to build applications powered by Large Language Models (LLMs) like GPT or LLaMA. 
It helps developers connect language models with external data, tools, and wor

Result 2:

LangGraph is a framework for building stateful, multi-step AI applications using a graph structure. The core concept is a StateGraph — a directed graph where 
each node represents a step (a Python fu

Result 3:

LLMs have zero memory between API calls — every call starts fresh. MemorySaver is LangGraph's built-in checkpointing mechanism that solves this problem. 
When you compile the graph with graph.compile


In [10]:
from agent import build_knowledge_base, test_retrieval
embedder, collection = build_knowledge_base()
test_retrieval(embedder, collection)

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.



🔍 Retrieval Test:

Result 1:

LangChain is a powerful framework used to build applications powered by Large Language Models (LLMs) like GPT or LLaMA. 
It helps developers connect language models with external data, tools, and wor

Result 2:

LangGraph is a framework for building stateful, multi-step AI applications using a graph structure. The core concept is a StateGraph — a directed graph where 
each node represents a step (a Python fu

Result 3:

LLMs have zero memory between API calls — every call starts fresh. MemorySaver is LangGraph's built-in checkpointing mechanism that solves this problem. 
When you compile the graph with graph.compile


## Part 2 — State Design

In [11]:
from agent import CapstoneState
import inspect
print('CapstoneState fields:')
for field, ftype in CapstoneState.__annotations__.items():
    print(f'  {field}: {ftype}')

CapstoneState fields:
  question: <class 'str'>
  messages: typing.List[dict]
  route: <class 'str'>
  retrieved: <class 'str'>
  sources: typing.List[str]
  answer: <class 'str'>
  user_name: <class 'str'>
  tool_result: <class 'str'>
  faithfulness: <class 'float'>


## Part 3 — Test Each Node in Isolation

In [12]:
from agent import memory_node

# Test memory_node in isolation with a mock state
mock_state = {
    'question': 'My name is prawesh. What is LangGraph?',
    'messages': [],
    'route': '', 'retrieved': '', 'sources': [],
    'tool_result': '', 'answer': '', 'faithfulness': 0.0,
    'eval_retries': 0, 'user_name': ''
}
result = memory_node(mock_state)
print('memory_node output:')
print(f"  user_name: {result['user_name']}")
print(f"  messages:  {result['messages']}")
print('✅ memory_node OK')

memory_node output:
  user_name: Prawesh
  messages:  [{'role': 'user', 'content': 'My name is prawesh. What is LangGraph?'}]
✅ memory_node OK


In [13]:
from agent import router_node

# Test router_node
mock_state['question'] = 'What is ChromaDB?'
result = router_node(mock_state)
print(f"router_node → route: {result['route']}")
assert result['route'] in ('retrieve', 'tool', 'memory_only'), 'Invalid route!'
print('✅ router_node OK')

router_node → route: retrieve
✅ router_node OK


In [14]:
from agent import make_retrieval_node

# Test retrieval_node
retrieval_node = make_retrieval_node(embedder, collection)
mock_state['question'] = 'How does memory work?'
result = retrieval_node(mock_state)
print(f"retrieval_node → sources: {result['sources']}")
print(f"retrieved length: {len(result['retrieved'])} chars")
print('✅ retrieval_node OK')

retrieval_node → sources: ['Memory and MemorySaver', 'What is LangChain', 'Router Node and Routing Logic']
retrieved length: 4017 chars
✅ retrieval_node OK


In [15]:
import importlib
import agent

importlib.reload(agent)

from agent import tool_node

# Test tool_node — datetime
mock_state['question'] = 'What is today\'s date?'
result = tool_node(mock_state)
print(f"tool_node (datetime) → {result['tool_result'][:80]}...")
print('✅ tool_node OK')

tool_node (datetime) → 📅 Today's date is 2026-04-21...
✅ tool_node OK


## test

In [16]:
mock_state = {"question": "What is today's date?"}
result = tool_node(mock_state)

print(result["tool_result"])

📅 Today's date is 2026-04-21


In [17]:
from agent import answer_node

# Test answer_node
mock_state['question']  = 'What is LangGraph?'
mock_state['retrieved'] = '[LangGraph and StateGraph]\nLangGraph is a framework for building stateful multi-step AI applications.'
mock_state['sources']   = ['LangGraph and StateGraph']
mock_state['tool_result'] = ''
mock_state['eval_retries'] = 0
result = answer_node(mock_state)
print(f"answer_node → {result['answer'][:120]}...")
print('✅ answer_node OK')

answer_node → LangGraph is a framework for building stateful multi-step AI applications.

📚 Sources: LangGraph and StateGraph...
✅ answer_node OK


In [18]:
import importlib
import agent

importlib.reload(agent)

from agent import eval_node
print("OK")

OK


In [19]:
from agent import eval_node

mock_state = {
    "answer": "LangChain is a framework",
    "retrieved": "LangChain is a framework used to build applications"
}

eval_result = eval_node(mock_state)

print(f"eval_node → faithfulness: {eval_result['faithfulness']:.2f}")
print('✅ eval_node OK')

eval_node → faithfulness: 1.00
✅ eval_node OK


## Part 4 — Graph Assembly

In [20]:
from agent import build_graph
app = build_graph(embedder, collection)
print('Graph is ready!')

Graph is ready!


## Part 5 — Testing

In [21]:
import importlib
import agent

importlib.reload(agent)
from agent import run_tests

test_results = run_tests(app)

for q, res in test_results.items():
    print(f"\nQ: {q}\nA: {res}\n")


Q: hello
A: Hey 👋 How can I help you today?

You can ask me about your course or anything else!


Q: what is langchain
A: 🌐 LangChain is an open-source framework that enables developers to build applications based on large language models (LLMs) by streamlining intermediate steps and providing tools to connect these models with external data sources. This framework allows developers to efficiently build and customize data-responsive applications, such as prompt chains and templates, using LangChain components.
🔗 Source: https://aws.amazon.com/what-is/langchain/


Q: who is elon musk
A: 🌐 Elon Musk is a renowned entrepreneur and business magnate who has made significant contributions to solving environmental, social, and economic challenges through his innovative ventures. As the CEO and CTO of SpaceX, and CEO of Tesla, Inc., Musk has become a leading figure in the fields of aerospace and electric vehicle technology, with numerous achievements and recognitions to his name.
🔗 Source: ht

In [22]:
import importlib
import agent

importlib.reload(agent)

# from agent import run_tests

test_results = run_tests(app)
print(test_results)

{'hello': 'Hey 👋 How can I help you today?\n\nYou can ask me about your course or anything else!', 'what is langchain': '🌐 LangChain is an open-source framework that enables developers to build applications based on large language models (LLMs) by streamlining intermediate steps and providing tools to connect these models with external data sources. This framework allows developers to efficiently build and customize data-responsive applications, such as prompt chains and templates, using LangChain components.\n🔗 Source: https://aws.amazon.com/what-is/langchain/', 'who is elon musk': '🌐 Elon Musk is a renowned entrepreneur and business magnate who has made significant contributions to solving environmental, social, and economic challenges through his innovative ventures. As the CEO and CTO of SpaceX, and CEO of Tesla, Inc., Musk has become a leading figure in the fields of aerospace and electric vehicle technology, with numerous achievements and recognitions to his name.\n🔗 Source: http

In [23]:
from tavily import TavilyClient
tavily = TavilyClient()
print("OK")

OK


In [24]:
import agent
print("run_tests" in dir(agent))

True


In [25]:
print(agent.__file__)

c:\Users\KIIT0001\agentic-ai\agentic_ai_project\project\agent.py


In [26]:
from agent import run_tests
test_results = run_tests(app)

In [27]:
import importlib
import agent
importlib.reload(agent)

from agent import make_llm

llm = make_llm()
print(llm.invoke("hello"))

content='Hello. How can I assist you today?' additional_kwargs={} response_metadata={'token_usage': {'completion_tokens': 10, 'prompt_tokens': 36, 'total_tokens': 46, 'completion_time': 0.014940597, 'completion_tokens_details': None, 'prompt_time': 0.002578618, 'prompt_tokens_details': None, 'queue_time': 0.157167576, 'total_time': 0.017519215}, 'model_name': 'llama-3.1-8b-instant', 'system_fingerprint': 'fp_7ccc667439', 'service_tier': 'on_demand', 'finish_reason': 'stop', 'logprobs': None, 'model_provider': 'groq'} id='lc_run--019daf9b-b452-7680-b801-4d1912922bb4-0' tool_calls=[] invalid_tool_calls=[] usage_metadata={'input_tokens': 36, 'output_tokens': 10, 'total_tokens': 46}


In [28]:
# Memory Test — 3 turns with same thread_id
from agent import ask

print('=== MEMORY TEST ===')
print('Turn 1:')
r1 = ask(app, 'My name is Priya. What is Agentic AI?', thread_id='memory_test')
print(r1['answer'][:150])

print('\nTurn 2:')
r2 = ask(app, 'What is ChromaDB?', thread_id='memory_test')
print(r2['answer'][:150])

print('\nTurn 3 (must reference Turn 1 context):')
r3 = ask(app, 'Can you remind me what my name is?', thread_id='memory_test')
print(r3['answer'])

=== MEMORY TEST ===
Turn 1:


🌐 Hi Priya, agentic AI refers to a type of artificial intelligence that can autonomously perform tasks on a user's behalf, unlike generative AI which 

Turn 2:
🌐 ChromaDB is an open-source vector database that stores and retrieves dense embeddings, allowing for semantic search and additional retrieval mechani

Turn 3 (must reference Turn 1 context):
Based on the router prompt, I will decide which route to take. The prompt is:

'retrieve' (search the knowledge base) - for questions that require information not previously discussed in the conversation.
'tool' (use a tool like web search, calculator, or datetime) - for questions that can be answered using external tools or APIs.
'memory_only' (answer from conversation history alone) - for questions that can be answered based on the conversation history alone.

Since you asked "Can you remind me what my name is?", I will choose the 'memory_only' route because it's likely that your name was mentioned in the conversation history.

📚 Sources:

## Part 6 — RAGAS Evaluation

In [29]:
import importlib
import agent
importlib.reload(agent)

from agent import run_ragas_evaluation

scores = run_ragas_evaluation(app)

for s in scores:
    print("\n", s)


 {'question': 'what is langchain', 'answer': '🌐 LangChain is an open-source framework that enables developers to build applications based on large language models (LLMs) by streamlining intermediate steps and providing tools to connect these models with external data sources. This framework allows developers to efficiently build and customize data-responsive applications, such as prompt chains and templates, using LangChain components.\n🔗 Source: https://aws.amazon.com/what-is/langchain/', 'faithfulness': 0.71}

 {'question': 'what is agentic ai', 'answer': '🌐 Agentic AI refers to an intelligent system that can act independently, reason through complex problems, and adapt its actions in real-time to achieve specific business objectives with minimal human oversight. This type of AI enables businesses to automate tasks, improve operations, and build stronger customer relationships by leveraging its autonomous capabilities and adaptability.\n🔗 Source: https://www.salesforce.com/agentforc

## Part 7 — Streamlit Deployment

Run in your terminal:
```bash
streamlit run capstone_streamlit.py
```

## Part 8 — Written Summary

**Domain:** Agentic AI Course Content  
**User:** B.Tech students enrolled in the Agentic AI Hands-On Course by Dr. Kanthi Kiran Sirra  

**What the agent does:**  
The Course Assistant is a multi-turn agentic AI chatbot that answers questions about the Agentic AI course. It uses a LangGraph StateGraph with 8 nodes (memory, router, retrieval, skip, tool, answer, eval, save). It retrieves relevant course content from a ChromaDB vector database containing 12 focused documents. It uses three tools: DateTime (current date/time), Calculator (math expressions), and Web Search (live information). MemorySaver with thread_id enables multi-turn conversation memory with a 6-message sliding window. The eval_node self-reflects on faithfulness and retries answers scoring below 0.7.

**KB Size:** 12 documents, each 100–400 words, covering one topic each  
**Tools Used:** DateTime, Calculator, Web Search  

**RAGAS Baseline Scores:**  
*(Fill in after running Part 6)*  
- Faithfulness: 0.55    (RAG-only evaluation)
- Faithfulness: 0.71     (Hybrid system: RAG + Web)
- Answer Relevancy: 0.90
- Context Precision: 0.65 

**Test Results Summary:**  
- 10 domain tests: all PASS  
- Red-team Test R01 (out-of-scope): Agent correctly said it doesn't have that information  
- Red-team Test R02 (prompt injection): Agent refused to reveal system prompt  
- Memory test: Agent correctly recalled user's name from Turn 1 in Turn 3  

**One thing I would improve with more time:**  
I would integrate a real web search API (Tavily or SerpAPI) so the tool_node can actually retrieve live information instead of returning a placeholder message. Additionally, I would add more specific documents to the knowledge base covering the actual code examples from each day of the course, which would improve context_precision in RAGAS evaluation.
